solutions/gold/usaco-696.mdx

```diff
@@ -306,10 +307,117 @@ int main() {
 </CPPSection>
 </LanguageSection>
 
-<IncompleteSection>
+## Solution 3: HLD
 
-Add Centroid Decomp/HLD Solution
+**Time Complexity:** $\mathcal{O}(N\log ^2N)$
+
+Sort the nodes from largest to smallest value, do path updates from each node to the root.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+class segTree {
+  private:
+	vector<int> segtree;
 
-Sort nodes from largest to smallest value, do path updates from itself the root.
+  public:
+	segTree(int N) {
+		if (__builtin_popcount(N) > 1) N = 1 << (32 - __builtin_clz(N));
```

Comment:
why would you need to do a `__builtin_popcount` check? idk seems a bit odd to me